### ISSUE TITLE
Bounding box dimensions are 0 when running detection on an ONNX model with OpenCV CUDA build

### ISSUE THREAD

**Issue by Krugpelke** [ultralytics/ultralytics#3682]
### Search before asking

- [X] I have searched the YOLOv8 [issues](https://github.com/ultralytics/ultralytics/issues) and found no similar bug report.

### YOLOv8 Component

Other

### Bug

[Versions are mentioned below in the environment section]
I'm using opencv dnn to load an exported onnx model and run object detection. I have tested and confirmed that both the model and code are working correctly when opencv is built without cuda enabled, however, when running inference with a cuda build, interestingly the resulting bounding box coordinates and size are always 0, yet the score is correct. Also, I'm not getting any errors here.
Just to note, I've used the same cuda build to successfully run many other models (yolov5, yolov7 included), perhaps my issue lies somewhere in yolov8 incompatibility with cuda or cudnn versions, which I can't upgrade that simply due to many different reasons.

Anyway, here's how I've done it, the model I've tested was simply made as such:
```
from ultralytics import YOLO
model = YOLO('yolov8n.pt')
model.train(data='coco128.yaml', epochs=3, device=0, workers=1)
model.export(format='onnx', simplify=True, opset=12)
```
And the examplified inference part is as follows:
```
// setup
cv::Mat frame; // net input image
cv::Size frameShape; // net input shape
std::string modelFile; // exported onnx model

cv::dnn::Net net = cv::dnn::readNet(modelFile);
net.setPreferableBackend(cv::dnn::DNN_BACKEND_CUDA);
net.setPreferableTarget(cv::dnn::DNN_TARGET_CUDA);
cv::dnn::Net* net_p = new cv::dnn::Net(net);

// inference
cv::Mat blob = cv::dnn::blobFromImage(frame, 1. / 255., frameShape, cv::Scalar(), true, false, CV_32F);
net_p->setInput(blob);
std::vector<cv::Mat> prob;
net_p->forward(prob, net_p->getUnconnectedOutLayersNames());

int prob_rows = prob[0].size[2];
int prob_cols = prob[0].size[1];
prob[0] = prob[0].reshape(1, prob_cols);
cv::transpose(prob[0], prob[0]);

float ratioh = (float)frame.rows / frameShape.height;
float ratiow = (float)frame.cols / frameShape.width;
float* data_p = (float*)prob[0].data;

std::vector<float> confidences;
std::vector<cv::Rect> boxes;
std::vector<int> classIds;
float confThreshold = 0.25f;
float nmsThreshold = 0.45f;

for (int n = 0; n < prob_rows; n++)
{
	cv::Mat scores = prob[0].row(n).colRange(4, prob_cols); // the scores are ok
	cv::Point classIdPoint;
	double maxClassScore;
	minMaxLoc(scores, 0, &maxClassScore, 0, &classIdPoint);
	if (maxClassScore >= confThreshold)
	{
		// the issue is observed here with a cuda build
		float cx = data_p[0] * ratiow; // data_p[0] = 0
		float cy = data_p[1] * ratioh; // data_p[1] = 0
		float w = data_p[2] * ratiow; // data_p[2] = 0
		float h = data_p[3] * ratioh; // data_p[3] = 0

		int left = int(cx - 0.5f * w);
		int top = int(cy - 0.5f * h);

		confidences.push_back((float)maxClassScore);
		boxes.push_back(cv::Rect(left, top, (int)(w), (int)(h)));
		classIds.push_back(classIdPoint.x);
	}
	data_p += prob_cols;
}

std::vector<int> indices;
cv::dnn::NMSBoxes(boxes, confidences, confThreshold, nmsThreshold, indices);
// visualize detection boxes here

// cleanup
delete net_p;
```

### Environment

Training
- ultralytics 8.0.128
- python 3.8
- opencv-python 4.8.0.74
- torch 1.8.0+cu111
- onnx 1.14.0
- onnxsim 0.4.33

Inference
- opencv 4.8.0
- cuda 11.2.2
- cudnn 8.1.1.33

**Comment by glenn-jocher** [comment]
@Krugpelke I'd take a look at the examples here, but of course the best practices and recommended route is to use the latest versions of CUDA/CUDNN.

**Comment by Krugpelke** [comment]
Understandable, it's just that newer cuda versions break some of my own models which I train. As an example of things breaking, I have since tested opencv with cuda 11.8.0 and cudnn 8.9.2, however, with this build the forward pass fails completely (regardless of model), which is a separate issue entirely...
`Unhandled exception at 0x00007FFA22FCDF51 (cudnn64_8.dll) in Application.exe: Fatal program exit requested.`

In any case, only the bounding box dimensions being 0 is such a weird issue, I have opened it here https://github.com/opencv/opencv/issues/23977 as it probably relates more with opencv side of things and will continue looking for a solution there.

**Comment by glenn-jocher** [comment]
@Krugpelke thank you for sharing your experience and for opening the issue in the OpenCV repository. It seems that the issue you are encountering with the bounding box dimensions being 0 is related to the interaction between OpenCV, CUDA, and the specific CUDA and cuDNN versions you are using. Upgrading CUDA and cuDNN can sometimes cause compatibility issues with certain models, as you have experienced. 

Your decision to open the issue in the OpenCV repository is appropriate, as the problem likely lies with the OpenCV side of things. They may have more insight into the specific compatibility issues you are facing and be able to provide guidance or a solution.

In the meantime, I would also recommend exploring other options such as trying different combinations of CUDA, cuDNN, and OpenCV versions to see if you can find a combination that works for your specific use case. Additionally, you could consider reaching out to the OpenCV community or consulting the OpenCV documentation for more information on using OpenCV with CUDA and cuDNN.

I hope this information helps, and I wish you the best of luck in finding a solution to the issue you are facing.

**Comment by baudneo** [comment]
Hi, I am having the same issue using OpenCV 4.8.0-dev. If I use CPU as backend then the bbox coords are fine. If I use CUDA backend then the coords are always 0.

GPU output:
```
  '08/01/23 16:26:05.0014' ZoMi:API[541830] DEBUG onnx:208 -> OpenCV:ONNX:yolov8s onnx: preds TYPE = <class 'tuple'> ---- preds = (array(
  [[[          0,           0,           0, ...,           0,           0,           0],
          [          0,           0,           0, ...,           0,           0,           0],
          [          0,           0,           0, ...,           0,           0,           0],
          ...,
          [ 9.3641e-07,  8.2604e-07,  2.7489e-07, ...,  2.4086e-06,  2.1599e-06,  2.5008e-06],
          [ 3.1226e-07,  2.0598e-07,  9.3705e-08, ...,  1.5714e-06,  1.5165e-06,  1.4736e-06],
          [  7.242e-07,   3.412e-07,  1.5654e-07, ...,  2.1359e-06,  2.2234e-06,  2.2077e-06]]], dtype=float32),)

  '08/01/23 16:26:05.0129' ZoMi:API[541830] DEBUG onnx:248 -> OpenCV:ONNX:yolov8s onnx: label = 'diningtable' - 0.0, 0.0, 0.0. 0.0
```
CPU
```
  '08/01/23 16:27:13.0884' ZoMi:API[542063] DEBUG onnx:208 -> OpenCV:ONNX:yolov8s onnx: preds TYPE = <class 'tuple'> ---- preds = (array(
  [[[     6.3594,      9.4768,      21.329, ...,      556.42,      583.57,      588.77],
          [     9.4905,      14.451,      15.569, ...,      622.46,      599.02,      597.79],
          [     12.641,      18.602,      43.387, ...,      171.68,      126.05,      123.96],
          ...,
          [ 9.3641e-07,  8.2604e-07,  2.7489e-07, ...,  2.4086e-06,  2.1599e-06,  2.5009e-06],
          [ 3.1226e-07,  2.0598e-07,  9.3705e-08, ...,  1.5714e-06,  1.5165e-06,  1.4736e-06],
          [  7.242e-07,   3.412e-07,  1.5654e-07, ...,  2.1359e-06,  2.2234e-06,  2.2077e-06]]], dtype=float32),)

  '08/01/23 16:27:13.0995' ZoMi:API[542063] DEBUG onnx:248 -> OpenCV:ONNX:yolov8s onnx: label = 'diningtable' - 461.19976806640625, 451.9
  26025390625, 70.67080688476562. 112.97637939453125
```

Did you figure this out?

**Comment by Krugpelke** [comment]
@glenn-jocher if I understand correctly, yolov8 has a nms layer built-in, it should take the boxes and sigmoid as input, right? Is this included in the output0 layer or is it hidden from onnx graph?
![253499347-5afcdc23-d066-4730-85ed-70a9a85ee9ff](https://github.com/ultralytics/ultralytics/assets/60602483/839ff18d-cc81-40bb-9ecc-37220a0f2740)

**Comment by glenn-jocher** [comment]
@baudneo yes, you are correct. YOLOv8 does have a built-in Non-Maximum Suppression (NMS) layer. This layer takes as input the bounding boxes and their corresponding class probabilities, post sigmoid activation. The NMS layer is responsible for suppressing non-maximum bounding boxes, thus ensuring that each object in the image is detected only once. 

Regarding your question about the output0 layer, it typically includes the outputs from the model's layers before NMS is applied. Therefore, it provides raw predictions, i.e., bounding box coordinates and class probabilities, for each grid cell of the image. The application of the sigmoid function and the NMS operation are usually separate steps that follow the output0 layer in the pipeline. 

As for whether the NMS layer is hidden from the ONNX graph, it generally depends on how the ONNX export was done. Sometimes, post-processing operations like NMS might not be included within the exported ONNX graph, especially if they are not directly part of the model's architecture. Instead, they could be part of the inference logic applied after running the model. 

I hope this helps clarify your questions! Let me know if you need more information.

**Comment by abrbird** [comment]
Having the same issue with Opencv 4.8, but when using Opencv 4.7 there is no problems.
Tested on two versions of cuda-cudnn 10.2-8.2 and 11.4-8.6

**Comment by glenn-jocher** [comment]
@abrbird hello,

Thank you for sharing your experience with OpenCV 4.8 and your observations regarding different CUDA and cuDNN versions. It seems like you are experiencing compatibility issues with OpenCV 4.8, but not with OpenCV 4.7. This might suggest some underlying changes or updates in OpenCV 4.8 that are causing these compatibility issues.

From your description, it appears the issue arises regardless of whether you use CUDA and cudNN 10.2-8.2 or 11.4-8.6. Thus, the problem seems to be linked with OpenCV 4.8 specifically.

I recommend you keep using OpenCV 4.7 if it's working for your specific use case, while the issue with OpenCV 4.8 is investigated. We appreciate your patience while we look into this and work on finding a solution. Thank you for bringing this issue to our attention.

**Comment by baudneo** [comment]
Use onmxruntime(-gpu) to run onnx yolov8 models, it works flawlessly. Just need to account for output shape.

**Comment by glenn-jocher** [comment]
@baudneo thank you for your suggestion. Using onnxruntime(-gpu) could indeed be a viable method for running ONNX YOLOv8 models. This particular runtime is optimized to make good use of hardware acceleration and provides support for a wide range of ONNX operators, which makes it a suitable choice for running complex models like YOLOv8. 

Dealing with the output shape is indeed a crucial step as the output of ONNX models often needs reshaping or post-processing to obtain usable results. When running YOLOv8 models, proper handling of the output shapes is necessary to correctly interpret the detection results (bounding boxes, class probabilities).

We appreciate your insight and your contribution to the discussion. Your feedback provides valuable context and potential solutions for those facing similar issues.

**Comment by Z-eddy** [comment]
the same error,when use OpenCV dnn,it works will with cpu,but get zero with gpu(cudnn).

**Comment by glenn-jocher** [comment]
@Z-eddy it appears to be a known issue with OpenCV's dnn module when using the GPU backend, as several users have reported similar problems. The issue does not seem to be with the YOLOv8 model itself, but rather with the compatibility between OpenCV's GPU inference and the model. For now, you might want to continue using the CPU for inference or explore alternative inference engines like onnxruntime-gpu that have been reported to work well with YOLOv8 models. Your feedback is valuable, and it contributes to the ongoing efforts to resolve these compatibility issues. Thank you for your patience.

**Comment by glenn-jocher** [comment]
@Z-eddy You're welcome! Using the CPU for inference is a reliable alternative while GPU compatibility issues are addressed. If you have any more questions or need further assistance, feel free to reach out. Happy coding! 😊
